plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,188 @@
+bindingUniqs ::
+    PLC.HasUnique name PLC.TermUnique =>
+    Binding tyname name uni fun (a, Uniques) ->
+    Uniques
+bindingUniqs b = Set.union (snd (bindingAnn b)) $ case b of
+    TermBind _a _s var _rhs -> Set.singleton $ var ^. PLC.varDeclName . PLC.unique
+    _                       -> mempty
+
+noUniq :: Functor f => f a -> f (a, Uniques)
+noUniq = fmap (,mempty)
+
+{- | The algebra for floating a `Binding` inwards. Given a `Term` and a `Binding`, determine
+ whether the `Binding` can be placed somewhere inside the `Term`.
+
+ If yes, return the result `Term`. Otherwise, return a `Let` constructed from
+ the given `Binding` and `Term`.
+-}
+floatAlg ::
+    forall tyname name uni fun a.
+    ( PLC.HasUnique name PLC.TermUnique
+    , PLC.ToBuiltinMeaning uni fun
+    ) =>
+    PLC.BuiltinVersion fun ->
+    -- | Annotation to be attached to the constructed `Let`.
+    a ->
+    Binding tyname name uni fun (a, Uniques) ->
+    Term tyname name uni fun (a, Uniques) ->
+    Term tyname name uni fun (a, Uniques)
+floatAlg ver letAnn = go
+  where
+    go ::
+        Binding tyname name uni fun (a, Uniques) ->
+        Term tyname name uni fun (a, Uniques) ->
+        Term tyname name uni fun (a, Uniques)
+    go b !body = case b of
+        TermBind _a Strict _var rhs
+            -- A strict binding whose RHS is not pure cannot be moved.
+            | not (isPure ver (const NonStrict) rhs) ->
+                let us = Set.union (termUniqs body) (bindingUniqs b)
+                 in Let (letAnn, us) NonRec (pure b) body
+        TermBind (_, usBind) _strictness var _rhs -> case body of
+            Apply (a, usBody) fun arg
+                | (var ^. PLC.unique) `Set.notMember` termUniqs fun ->
+                    -- `fun` does not mention `var`, so we can place the binding inside `arg`.
+                    Apply (a, Set.union usBind usBody) fun (go b arg)
+                | (var ^. PLC.unique) `Set.notMember` termUniqs arg ->
+                    -- `arg` does not mention `var`, so we can place the binding inside `fun`.
+                    Apply (a, Set.union usBind usBody) (go b fun) arg
+            LamAbs (a, usBody) n ty lamAbsBody ->
+                -- A binding can always be placed inside a `LamAbs`.
+                LamAbs (a, Set.union usBind usBody) n ty (go b lamAbsBody)
+            TyAbs (a, usBody) n k tyAbsBody ->
+                -- A binding can always be placed inside a `TyAbs`.
+                TyAbs (a, Set.union usBind usBody) n k (go b tyAbsBody)
+            TyInst (a, usBody) tyInstBody ty ->
+                -- A binding can always be placed inside a `TyInst`.
+                TyInst (a, Set.union usBind usBody) (go b tyInstBody) ty
+            Let (a, usBody) r bs letBody
+                -- The binding can be placed inside a `Let`, if the right hand sides of the
+                -- bindings of the `Let` do not mention `var`.
```

Comment:
What about pushing it into one of the bindings? i.e.
```
let x = 1
in 
let 
   y = x 
   z = 1
in ...
```
Would be nice to turn that into
```
let 
  y = let x = 1 in x
  z = 1
in ...
```

Reply:
Hmm, good point, I didn't think of this!

Reply:
Well, I tried it (https://github.com/input-output-hk/plutus/commit/c88f5b088da588caaab684fd26c42e746bced05b), but it makes things worse. It's hard to tell exactly why since the pir/plc are quite big, but my guess is that once we turn `let a = ... in let b = a in body` into `let b = (let a = ... in a) in body`, the cost and size of `b` are no longer acceptable, which affects inlining.

Reply:
I mean, I think we can do the obviously sensible floating in behaviour here and worry about making the inliner better separately.

And there's really no reason not to add an indicative test case.

Reply:
@michaelpj I think I know why - it has nothing to do with inlining. It's because of floating bindings into lambdas. We only float lazy or strict pure bindings, so when floating into a lambda, the RHS itself is not evaluated more times. However, because a binding is compiled into `(\x -> body) (delay rhs)`, it incurs more machine steps: Delay, Apply and LamAbs. This is confirmed by counting the machine steps - only these three steps increased.

The `checkScriptContext` test cases have `fmap mkTxOut [1..i]`, so this all makes total sense now. We should avoid floating bindings into lambdas.

Reply:
Right, so there's some overhead from the let-binding itself, that's a good point. But that seems like a problem which is _independent_ of what we're discussing in this thread, right? Which is about floating things into the RHS of let-bindings. So I'm not sure that explains why we get an increase, unless I'm missing something.

> Delay, Apply and LamAbs. This is confirmed by counting the machine steps - only these three steps increased.

Most of the program is applications and lambda abstractions! I'm not sure how much this tells us...

Reply:
> But that seems like a problem which is independent of what we're discussing in this thread, right? Which is about floating things into the RHS of let-bindings. So I'm not sure that explains why we get an increase, unless I'm missing something.

This is exactly why! The RHS of let-bindings contains lambdas, and by floating things into RHS, we float them into lambdas.

Another way to demonstrate this is

> fmap mkTxOut [1..i]

When `i=0`, floating things into RHS saves cost. When `i=1`, it breaks even. The larger `i` is, the worse it gets.

Reply:
> Most of the program is applications and lambda abstractions! I'm not sure how much this tells us...

Btw, it can also be observed that when floating things into RHS, `Delay`, `Apply` and `LamAbs` increase by the same number, and that number equals `i-1`. The numbers of all other machine steps remain the same.

Reply:
Ah, so indeed the correct fix is to be more careful about floating into lambdas. That makes sense why you wouldn't have seen the problem before - indeed most lambdas probably appear on the RHS of bindings.

Well, I feel vindicated for claiming that the decision of whether or not to float into lambdas is complicated and needs a careful comment at least :)

Reply:
I'm confused. I think some examples would really really help (as test cases!).

I actually think that what I said was wrong. It's not _lambdas_ that cause multiple evaluation, it's _applications_. The only way that something can get evaluated multiple times is by being bound to a variable and then having the variable used multiple times, i.e. by being in the _argument_ of an application.

e.g. 

```
let x = ...
in [(\y . ... y... y...) x]
```
x gets evaluated multiple times, despite not being in a lambda.

And of course the RHSs of bindings _are_ arguments of applications and indeed can lead to things being evaluated more often. If you pushed a let-binding into the RHS of the x binding above it would get evaluated twice.

So I think the strict rule is:
- Don't float into the arguments of applications
- Don't float into the RHS of bindings

Maybe we can relax that? I'm not sure. But this needs a careful note and some examples.

Reply:
Hmm, I have to think about it a bit more. I thought lambdas with a bigger body is obviously more expensive if applied multiple times, but I guess it's not that simple?

Reply:
Well, it's not about how expensive `expensive` is. The `expensive` itself will always be evaluated the same number of times (as long as `x` is lazy; if `x` is strict, we wouldn't move `let x = expensive`). It's about how many times `delay expensive` is evaluated. Each time `delay expensive` is evaluated, it incurs an additional LamAbs, Apply and Delay step.

So how many times is `delay expensive` evaluated in this example? The compiled code for the three cases are, respectively
```
(\x -> (\f -> f1 f2) (\y -> ..x..)) (delay expensive)
(\f -> f1 f2) ((\x -> (\y -> ..x..)) (delay expensive))
(\f -> f1 f2) (\y -> (\x -> ..x..) (delay expensive))
```

As far as I can tell, case 1 evaluates `delay expensive` once, while cases 2 and 3 evaluate it twice. So I think what you said previously:

> So I think the strict rule is:
Don't float into the arguments of applications
Don't float into the RHS of bindings

is correct.

Reply:
Wait, I was wrong - case 2 also evaluates `delay expensive` only once. I wrote and ran this test, and 1 and 2 have the same cost, and 3 has a higher cost. So, it should probably be "Don't float into lambdas within RHSs of bindings" and "Don't float into lambdas within arguments of applications"...

Reply:
I'm going to add a small note of caution here.  From my experiments with Plutonomy (see the notes attached to PLT-1044), it seemed that the float-in optimisation was most effective in combination with some of the other optimisations.  Rather than trying to make this particular transformation as good as possible at this stage, it might be sensible to implement something simple and then add other optimisations as separate steps in the pipeline.  In the long run that might amount to the same thing as having one big super-efficient inliner, but it could be conceptually simpler if we have a number of separate stages.  I'm just guessing though: I could easily be wrong!

Also, I think part of the value of this is understanding _why_ all this stuff works, and I think we're beginning to see that in some of the discussion here.  We should try to make a record of the thinking that's been going on...

Reply:
@michaelpj Ok, I've tried a number of alternatives on our existing tests, including
- no floating into any lambda abstractions
- no floating into any arguments
- no floating into lambda abstractions that are in RHSs of bindings
- no floating into lambda abstractions that are in RHSs of bindings or in arguments

And these are all _strictly worse_ than what this PR currently does (i.e., no floating into RHSs). Here's my explanation:
- Not floating into arguments at all is bad, because in an argument there may be unevaluated branches, and floating bindings into those branches saves cost.
- Not floating into lambdas at all is bad, for similar reasons. Also many lambdas may not be applied at all.
- Floating into lambdas inside RHSs can be bad, as demonstrated in Michael's example above (case 3).
- Theoretically, if floating into RHSs can be bad, then floating into arguments can also be bad, because `let x = RHS in body` is the same as `(\x -> body) RHS`. However, due to GHC's transformations we get the former much more often than the latter (even if we don't run our own beta-reduction pass).
- Floating into RHSs, even without floating into lambdas, can _also be bad_! The following is an example:

```haskell
let x=1+2
 in let y=x+x
     in y+y
```

Without floating, this is turned into
```
(\x -> (\y -> y+y) (delay $ x+x)) (delay 1+2)
```

and `delay ...` is evaluated twice. If `x` is floated into the RHS of `y`, then it is turned into

```
(\y -> y+y) (delay $ (\x -> x+x) (delay 1+2))
```

and `delay ...` is evaluated three times (note the second `delay` is inside the first `delay`)!

Based on this observation, I think what this PR currently does is what we should do. @kwxm Hopefully this is a good thread to refer to in the future!

Reply:
> Theoretically, if floating into RHSs can be bad, then floating into arguments can also be bad, because let x = RHS in body is the same as (\x -> body) RHS. However, due to GHC's transformations we get the former much more often than the latter (even if we don't run our own beta-reduction pass)

I don't find this convincing. These two _are_ very similar, and your argument for floating into arguments "there might be unevaluated branches" is just as valid for let-RHSs.

> Without floating, this is turned into

To be clear, you're talking about non-strict let-bindings here (we should _really_ be explicit, it makes a big difference!). Strict let-bindings don't have the delays, and are quite different.

----

One thing that I think we're looking for the answer to the question of "when is a let binding work-free?". I think the answer is that _no let-binding is ever work-free_ because of the abstraction/application overhead of processing the binding. That's an important thing to note, IMO it means it is _never_ fully safe to float _any_ let-binding into a context where we can repeat work.

We might want to relax this in some cases, since the overhead is small, but I think it's useful to note that strictly they're never work-free.

----

> Floating into RHSs, even without floating into lambdas, can also be bad! The following is an example:

I think what you're pointing to again is about _delaying evaluation_ anything that delays evaluation, which is lambdas, type abstractions, _and the rhs of non-strict bindings_ (I was missing this one!) is problematic and can lead to repeated evaluation.

I would buy a heuristic that says not to push anything that's not work-free into anything that can lead to repeated evaluation. I think you want to carve out some special cases but I'm not sure that what you've written is principled enough.

I think there is a variant of this that is principled, though. Given a term, I think we can identify the set of child nodes that will be evaluated _strictly the same number of times_ as the overall term. We do that by recursing into only the bodies of let-terms and the _functions_ (not arguments!) of applications. Then it _is_ safe to move even non-work-free bindings within this subset of nodes, since it'll still be evaluated strictly the same number of times.

-----

I don't think this is complete, but I think it's getting there. Perhaps for the first version we should skip the equally-evaluated thing and just not float into things that can delay evaluation.

-----

> Hopefully this is a good thread to refer to in the future!

THIS MUST GO INTO A NOTE!! I'm sure you know that but I just really want to emphasize this. Looking back for an old PR comment thread is _not_ an adequate way of recording this sort of knowledge! We really want the fully-processed version of this discussion recorded alongside the pass.

Reply:
> I would buy a heuristic that says not to push anything that's not work-free into anything that can lead to repeated evaluation. I think you want to carve out some special cases but I'm not sure that what you've written is principled enough.

Unfortunately, if we avoid floating into any RHS or any argument whatsoever, then there's zero savings among our test cases.

Although a lazy binding or a strict pure binding, I think it is unlikely that floating them inwards leads to unbounded cost increase. The additional cost from one more LamAbs, Apply and Delay is likely small compared to whatever real work is being done.

So unless there's a better approach that guarantees no cost increase whatsoever while preserving at least some benefits, we'd have to make a decision between not floating-in at all vs. accepting some small cost increases.

Reply:
> Although a lazy binding or a strict pure binding is not work-free, I think it is unlikely that floating them inwards leads to unbounded cost increase.

Yeah, I think this would also be reasonable. We just need an `isWorkFreeApartFromOverhead` predicate and a clear explanation. But it does mean that there probably will be _some_ cases where things get worse.

Reply:
> We just need an isWorkFreeApartFromOverhead predicate and a clear explanation

That's just `isPure`, I think, which means it's the same as what I'm currently doing, or are you thinking about something else?

Reply:
`isPure` is a property of terms, `isWorkFreeWithoutOverhead` is a property of _bindings_, since the overhead comes from the binding. I just think it's good to have explicit names for the concepts we're using, even if they're ultimately implemented very similarly.